[cxxmodule] Enable explicit modules

yamaguchi1024 opened the pull request:
These changes includes adding `--cxxmodule` flag to genreflex, and
adding interface headers to classes.h

We are planning to automatically generate them
(https://github.com/cms-sw/cmssw/issues/25903), but for now we can
manually maintain them.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25916/8397

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @yamaguchi1024 (Yuka Takahashi) for master.

It involves the following packages:

DataFormats/Common
DataFormats/DetId
DataFormats/GeometrySurface
DataFormats/GeometryVector
DataFormats/Math
DataFormats/Provenance
DataFormats/StdDictionaries
DataFormats/TrajectoryState
FWCore/MessageLogger


@perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @wddgit, @VinInn, @Martin-Grunewald, @rovere, @gpetruc this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

smuzaffar commented:
@yamaguchi1024 , I would not suggest to test/merge this in cmssw master. No idea how root 6.12 , 6.14 and master IBs will behave with --cxxmodule flag.

yamaguchi1024 commented:
@smuzaffar , They will ignore the --cxxmodule flag, but I forgot to add <if CXXMODULE>, I will update this

cmsbuild commented:
-1

Tested at: 51a795b26fabb573ad170ad91b29b8a26bb5f953

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25916/33096/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found compilation error when building: 

<pre>Entering library rule at DataFormats/GeometryVector
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometryVector/src/TrackingJacobians.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometryVector/src/print.cc 
>> Building LCG reflex dict from header file src/DataFormats/GeometryVector/src/classes.h
In file included from input_line_9:64:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometryVector/interface/extBasic2DVector.h:15:7: error: redefinition of 'Basic2DVector'
class Basic2DVector {
      ^
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometryVector/interface/oldBasic2DVector.h:17:7: note: previous definition is here
class Basic2DVector {
      ^
</pre>

* **Clang**:

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 16 COMPILER='llvm compile'

```


<pre>>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometrySurface/src/RectangularPlaneBounds.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometrySurface/src/BoundDisk.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometrySurface/src/Plane.cc 
>> Building LCG reflex dict from header file src/DataFormats/GeometrySurface/src/classes.h
In file included from input_line_9:52:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometrySurface/interface/extTkRotation.h:32:7: error: redefinition of 'TkRotation'
class TkRotation {
      ^
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-12-1100/src/DataFormats/GeometrySurface/interface/oldTkRotation.h:32:7: note: previous definition is here
class TkRotation {
      ^
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

slava77 commented:
-1

resolution(s) to the build errors and other update are expected, apparently

yamaguchi1024 committed:
[cxxmodule] Enable explicit c++ modules

This PR modularizes 25 libraries, together with
https://github.com/cms-sw/cmssw-modulemap.

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented on DataFormats/Common/src/classes.h:
```diff
@@ -1,34 +1,118 @@
-#include "DataFormats/Common/interface/ConstPtrCache.h"
-#include "DataFormats/Common/interface/BoolCache.h"
-#include "DataFormats/Common/interface/PtrVectorBase.h"
+#include "DataFormats/Common/interface/ValidHandle.h"
```
There is absolutely no reason to make a ROOT dictionary for ValidHandle. If `classes.h` is intended just for dictionary generation, putting all headers into the file doesn't seem like a good idea. If the intent is preparation for C++ modules, then, OK, I guess.

yamaguchi1024 replied:
I see. It is a preparation for C++ modules, but why ValidHandle.h is in interface/ if it is not supposed to be exposed to users? (Shouldn't it be under src if it was like an internal header?)

Dr15Jones replied:
Up until now, we've only used classes.h to create the ROOT dictionaries we need for either I/O or for access to member functions of certain classes. `ValidHandle` is not necessary for either of those cases but it is needed for other code in CMSSW.

If classes.h is now being used as preparation for C++ modules then `ValidHandle` should be there.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25916/8497

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #25916 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @cmsbuild, @slava77, @santocch can you please check and sign again.

Dr15Jones commented:
@smuzaffar how should this be tested? Just running the regular tests?

smuzaffar commented:
@Dr15Jones , I have changed this PR to be based on 10.6.CXXMODULE.X branch. This should be testing against CXXMODULE IB to make use of different build rules.

smuzaffar commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

smuzaffar commented on DataFormats/Math/interface/AVXVec.h:
```diff
@@ -1,6 +1,8 @@
 #ifndef DataFormat_Math_AVXVec_H
 #define DataFormat_Math_AVXVec_H
 
+#include "DataFormats/Math/interface/SSEVec.h"
```
@yamaguchi1024 , is not it going to create a cyclic dependency (SSEVec.h includes AVXVec.h too)?

yamaguchi1024 replied:
It'll not be a cyclic dependency if headers were in the same module (SSWVec.h and AVXVec.h are both in DataFormatsMath_xr.pcm), but cyclic dependency occurs when headers in the different module are including each other. (Like https://github.com/cms-sw/cmssw/issues/25903#issuecomment-463298741 case)

smuzaffar commented:
@yamaguchi1024 , CXXMODULE IB build rules already add all missing packages/interface/*.h in to its src/classes.h. Do you still explicitly wants to add them?

cmsbuild commented:
-1

Tested at: d02faad9264af6cee484b475422ee6984ed6ead5

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25916/33223/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency**

* **Build**:

I found compilation error when building: 

<pre>>> Entering Package DataFormats/StdDictionaries
>> Building LCG reflex dict from header file src/DataFormats/StdDictionaries/src/classes_vector.h
>> Building LCG reflex dict from header file src/DataFormats/StdDictionaries/src/classes_map.h
>> Building LCG reflex dict from header file src/DataFormats/StdDictionaries/src/classes_pair.h
>> Building LCG reflex dict from header file src/DataFormats/StdDictionaries/src/classes_others.h
./modulemap/module.modulemap:1:8: error: redefinition of module 'DataFormatsCommon_xr'
./modulemap/module.modulemap:1:8: error: module "DataFormatsCommon_xr" {redefinition of module 'DataFormatsCommon_xr'
./modulemap/module.modulemap
       ^
:1:8: error: module "DataFormatsCommon_xr" {
redefinition of module 'DataFormatsCommon_xr'       ^
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

yamaguchi1024 commented:
This error is because Clang is finding two modulemaps defining the same module in its INCLUDE_PATH (`/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_CXXMODULE_X_2019-02-21-2300/src/module.modulemap` and `./module.modulemap`). I feel IB's paths are not intended to be added to INCLUDE_PATH at build time (If someone accidentally delete one header in the current build but the compiler still finds the old header from IB paths?), but what do you think?

yamaguchi1024 commented:
> @yamaguchi1024 , CXXMODULE IB build rules already add all missing packages/interface/*.h in to its src/classes.h. Do you still explicitly wants to add them?

Oh great! Then I guess we don't need classes.h modifications.

slava77 commented:
@yamaguchi1024 
please clarify on the status/plans for this PR.
Thank you.

yamaguchi1024 commented:
@slava77 Thanks for the heads-up!

Closing PR as it is outdated, we have now auto-generated module.modulemap and classes.h thanks to @smuzaffar . From now when we want to enable C++ Modules for CMS libraries, all we need to do is to add
```
<release name="_CXXMODULE_"><flags   GENREFLEX_ARGS="--deep --cxxmodule"/></release>
```
to BuildFile.xml.
